modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/MatrixStatsResults.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.matrix.stats;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Descriptive stats gathered per shard. Coordinating node computes final pearson product coefficient
+ * based on these descriptive stats
+ *
+ * @internal
+ */
+class MatrixStatsResults implements Streamable {
+    /** object holding results - computes results in place */
+    final RunningStats results;
+    /** pearson product correlation coefficients */
+    protected HashMap<String, HashMap<String, Double>> correlation;
+
+    /** Base ctor */
+    private MatrixStatsResults() {
+        results = RunningStats.EMPTY();
+        this.correlation = new HashMap<>();
+    }
+
+    /** creates and computes result from provided stats */
+    public MatrixStatsResults(RunningStats stats) {
+        try {
+            this.results = stats.clone();
+            this.correlation = new HashMap<>();
+        } catch (CloneNotSupportedException e) {
+            throw new ElasticsearchException("Error trying to create multifield_stats results", e);
+        }
+        this.compute();
+    }
+
+    /** creates a results object from the given stream */
+    protected MatrixStatsResults(StreamInput in) {
+        try {
+            results = new RunningStats(in);
+            this.readFrom(in);
```

Comment:
Can't remove `readFrom()` since its part of the Interface. Do you want me to just noop?